### ISSUE TITLE
How to install and use your packages?

### ISSUE THREAD

**Issue by Velmurugan1008** [Johnny1188/fractional-learning-to-optimize#1]
I tried to use your packages in my Python environment, I can't import your packaged in my code. Please tell us the procedure for installing your package in our Python environment.

**Comment by Johnny1188** [comment]
Hello @Velmurugan1008, this repository is currently a work in progress and should be fully ready to use in the next month or two (README file, dependencies, more changes, etc.). If you would like to use the code now, `pip install -e .` in the root folder will install the package `fl2o` itself in an editable mode. I don't have access to the environment at the moment, so I cannot tell you the exact dependencies, but you can run it and see what's missing in your particular environment. From the top of my head, I can tell you that I've been running it with Python 3.10 and PyTorch 2.X.

**Comment by Velmurugan1008** [comment]
Thank you for your reply. I want to rewrite your class FGD by modifying the fractional gradient based on the Caputo L1 scheme. See equation (4.4) in the attached PDF. Here is the modified code (please check my code, and if there are any errors or corrections, please let me know):

class FGD(nn.Module):
    """
    Fractional (Stochastic) Gradient Descent with Higher Order Truncation (hot) and/or Fixed Memory Step (K>=1)
    """

    def __init__(
        self,
        params,
        lr=0.01,
        alpha=1,
        c=0,
        eps=1e-8,
        hot=True,
        K=0,
        device="cpu",
    ):
        assert hot, "Only hot=True is supported for now."
        super().__init__()

        self.param_groups = [
            {"params": None, "lr": lr, "alpha": alpha, "c": c, "eps": eps, "hot": hot, "K": K, "device": device}
        ]
        self.state = []
        for p in params:
            p.retain_grad()
            self.state.append({})
        self.hot = hot
        self.K = K
        self.device = device

    def zero_grad(self):
        for g in self.param_groups:
            for p in g["params"]:
                if p.grad is not None:
                    p.grad.zero_()

    def step(self, optee):
        for g in self.param_groups:
            updated_params = {}
            for p_idx, (n, p) in enumerate(optee.named_parameters()):
                if p.grad is None:
                    continue
                if len(self.state[p_idx]) == 0:
                    self.state[p_idx]["step"] = 1
                    self.state[p_idx]["memory"] = [g["c"] for _ in range(max(1, g["K"]))]

                grad = p.grad.data
                h = 0.0
                for j in range(len(self.state[p_idx]["memory"])-1):
                    
                    tmp = ((p.data.detach() - self.state[p_idx]["memory"][j])**(1 - g["alpha"]) - (p.data.detach() - self.state[p_idx]["memory"][j+1])**(1 - g["alpha"]))
                    gam = gamma(2 - g["alpha"])
                    h += grad * tmp / gam

                updated_params[n] = p.data - g["lr"] * h

                if g["K"] >= 1:
                    # self.state[p_idx]["memory"].pop(0)
                    self.state[p_idx]["memory"].append(p.data.detach().clone())
                self.state[p_idx]["step"] += 1

            for n, p in optee.named_parameters():
                p.data = updated_params[n]

        return None
[chapter-4-numerical-caputo-differentiation-2019.pdf](https://github.com/user-attachments/files/16103956/chapter-4-numerical-caputo-differentiation-2019.pdf)
<img width="583" alt="Algorithm_FGD" src="https://github.com/Johnny1188/fractional-learning-to-optimize/assets/117431820/4e991c1e-e2ec-4946-9b66-a01cba92bb3f">

Your message is more helpful to me. Thank you!

**Comment by Johnny1188** [comment]
All seem fine except for a few things:
1) In Algorithm 2 that you attached, the for-loop starting on line 4 mixes the index $j$ with the inner summation. Moreover, I don't see any use of the for-loop (and it is not in your code), could you explain?
2) In the Algorithm 2, the first-order derivative of $f$ is evaluated at past $x_j$, but in your code, you are using the first-order derivative (the gradient) evaluated at the current $x_k$ (line `h += grad * tmp / gam`).
3) Are you going to use the learned optimizer together with the FGD? If so, keep in mind that it sets the $\alpha$ parameter as pytorch tensor (to be able to backpropagate through it back to the learned optimizer's parameters). So, for example, `gamma(2 - g["alpha"])` will not work in this combination because scipy's `gamma` function doesn't work with pytorch tensors.
4) If you expect that all parameters are going to have the `.grad` filled, consider removing `if p.grad is None: continue` to forego any hidden bugs.
5) The `gamma` function isn't cheap to evaluate; consider moving `gam = gamma(2 - g["alpha"])` outside the for-loop.

**Comment by Velmurugan1008** [comment]
1. Now, I rewrite the algorithm as follows:

<img width="411" alt="Screenshot 2024-07-09 130349" src="https://github.com/Johnny1188/fractional-learning-to-optimize/assets/117431820/4e6a3f67-73e8-4828-bb8d-190ddef9516e">

2. I want to store the values of grad in each step and use them back in the for loop. Please suggest how to store the grad values, like the parameters stored as memory in your FGD code.
3.  As per your suggestion, I defined the `gamma` outside the` for loop`  and also defined `gamma` in terms of `tensor`

I want to write Python optimization code based on your class FGD. Here is the modified code (please check my code, and if there are any errors or corrections, please let me know):

def __init__(
    self,
    params,
    lr=0.01,
    alpha=1,
    c=0,
    eps=1e-8,
    hot=True,
    K=2,
    device="cpu",
):
    assert hot, "Only hot=True is supported for now."
    super().__init__()

    self.param_groups = [
        {"params": None, "lr": lr, "alpha": alpha, "c": c, "eps": eps, "hot": hot, "K": K, "device": device}
    ]
    self.state = []
    for p in params:
        p.retain_grad()
        self.state.append({})
    self.hot = hot
    self.K = K
    self.device = device

def zero_grad(self):
    for g in self.param_groups:
        for p in g["params"]:
            if p.grad is not None:
                p.grad.zero_()

def step(self, optee):
    for g in self.param_groups:
        updated_params = {}
        for p_idx, (n, p) in enumerate(optee.named_parameters()):
            if p.grad is None:
                continue
            if len(self.state[p_idx]) == 0:
                self.state[p_idx]["step"] = 1
                self.state[p_idx]["memory"] = [g["c"] for _ in range(max(1, g["K"]))]

            grad = p.grad.data
            gam = torch.exp(torch.lgamma(torch.tensor(2 - g["alpha"])))
            h = 0.0
            for j in range(len(self.state[p_idx]["memory"])-1):
                
                tmp = ((p.data.detach() - self.state[p_idx]["memory"][j])**(1 - g["alpha"]) - (p.data.detach() - self.state[p_idx]["memory"][j+1])**(1 - g["alpha"]))
               
                h += grad * tmp / gam

            updated_params[n] = p.data - g["lr"] * h

            if g["K"] >= 1:
                # self.state[p_idx]["memory"].pop(0)
                self.state[p_idx]["memory"].append(p.data.detach().clone())
            self.state[p_idx]["step"] += 1

        for n, p in optee.named_parameters():
            p.data = updated_params[n]

    return None
 Your message is more helpful to me. Thank you!

**Comment by Johnny1188** [comment]
1. Your rewritten algorithm seems fine now. But I noticed that you changed where the first derivative is evaluated inside the summation. Namely, before, you had an evaluation at $x_j$, and now it's $x_{k+1}$. I don't know what should be there, but I just wanted you to know...
2. You can store the gradients from each step in a list similarly to the parameters themselves (see `self.state[p_idx]["memory_grad"]` below). But you will need to evaluate the gradient for each of the $x_0, \dots, x_{K+1}$ and pass them to the optimizer (see the new class parameter `c_grad`; the default value most probably doesn't match with the `c` values so keep that in mind).
3. In the case where `g["alpha"]` is pytorch tensor, `torch.tensor(2 - g["alpha"])` would make a copy, and I'm not completely sure that it would correctly continue the computational graph and not break meta-training of the learned optimizer (which sets the `g["alpha"]`). In any case, copying like this might slow it down a bit, so I would rather go with the solution below (as in the original code).

```python
def init(
    self,
    params,
    lr=0.01,
    alpha=1,
    c=0,
    c_grad=0,
    eps=1e-8,
    hot=True,
    K=2,
    device="cpu",
):
    assert hot, "Only hot=True is supported for now."
    super().init()

    self.param_groups = [
        {"params": None, "lr": lr, "alpha": alpha, "c": c, "c_grad": c_grad, "eps": eps, "hot": hot, "K": K, "device": device}
    ]
    self.state = []
    for p in params:
        p.retain_grad()
        self.state.append({})
    self.hot = hot
    self.K = K
    self.device = device

def zero_grad(self):
    for g in self.param_groups:
        for p in g["params"]:
            if p.grad is not None:
                p.grad.zero_()

def step(self, optee):
    for g in self.param_groups:
        updated_params = {}
        for p_idx, (n, p) in enumerate(optee.named_parameters()):
            if p.grad is None:
                continue
            if len(self.state[p_idx]) == 0:
                self.state[p_idx]["step"] = 1
                self.state[p_idx]["memory"] = [g["c"] for _ in range(max(1, g["K"]))]
                self.state[p_idx]["memory_grad"] = [g["c_grad"] for _ in range(max(1, g["K"]))]

            grad = p.grad.data
            if type(g["alpha"]) in (float, int):
                gam = gamma(2 - g["alpha"])
            elif type(g["alpha"]) == torch.Tensor:
                gam = torch.lgamma(2 - g["alpha"]).exp()
            else:
                raise Exception("This should not happen.")
            h = 0.0
            for j in range(len(self.state[p_idx]["memory"])-1):
                tmp = ((p.data.detach() - self.state[p_idx]["memory"][j])**(1 - g["alpha"]) - (p.data.detach() - self.state[p_idx]["memory"][j+1])**(1 - g["alpha"]))
                h += self.state[p_idx]["memory_grad"][j+1] * tmp / gam

            updated_params[n] = p.data - g["lr"] * h

            if g["K"] >= 1:
                self.state[p_idx]["memory"].append(p.data.detach().clone())
                self.state[p_idx]["memory_grad"].append(grad.detach().clone())
            self.state[p_idx]["step"] += 1

    for n, p in optee.named_parameters():
        p.data = updated_params[n]

    return None
```

**Comment by Velmurugan1008** [comment]
Thank you for your message! It will be helpful for me.
